### PR TITLE
MDS-3646: Users Can Create Work Information (Start/Stop) Records in Core

### PR DESCRIPTION
**Pull request by exzizt**
# Main
* Users can now create/edit/delete "work information" records in Core. These records will contain the work "start/stop" data as well as other data in the future.
image

image

image

* Users can see and filter by the mine work status on the mine view.
image

# Other
* Changes the order of the columns in the Browse Mines view

### CODE REVIEW

File: services/core-web/common/actionCreators/incidentActionCreator.js [hunk 1]
```diff
@@ -80,24 +80,24 @@ export const fetchIncidents = (payload) => (dispatch) => {
...
 
-export const deleteMineIncident = (mineGuid, incidentGuid) => (dispatch) => {
-  dispatch(request(reducerTypes.DELETE_INCIDENT));
+export const deleteMineIncident = (mineGuid, mineIncidentGuid) => (dispatch) => {
+  dispatch(request(reducerTypes.DELETE_MINE_INCIDENT));
   dispatch(showLoading());
   return CustomAxios()
     .delete(
-      `${ENVIRONMENT.apiUrl}${API.INCIDENT_DELETE(mineGuid, incidentGuid)}`,
+      `${ENVIRONMENT.apiUrl}${API.MINE_INCIDENT(mineGuid, mineIncidentGuid)}`,
```
Comment (exzizt) [inline services/core-web/common/actionCreators/incidentActionCreator.js:88]: Tidied up the incident action creator.

File: migrations/sql/V2021.05.25.15.10__create_mine_work_information_table.sql [hunk 2]
```diff
@@ -0,0 +1,50 @@
+CREATE TABLE IF NOT EXISTS mine_work_status (
+  mine_work_status_code varchar PRIMARY KEY,
```
Comment (mhuseinov) [inline migrations/sql/V2021.05.25.15.10__create_mine_work_information_table.sql:2]: Usually, our code-tables have PK of varchar(3), is there any reason to have these without any limit?

File: services/core-api/app/api/mines/mine/models/mine.py [hunk 3]
```diff
@@ -162,6 +171,27 @@ def has_minespace_users(self):
...
+    def mine_work_information(self):
+        if self.mine_work_informations:
+            return self.mine_work_informations[0]
+        return None
+
+    @hybrid_property
+    def work_status(self):
+        if self.mine_work_informations:
+            return self.mine_work_informations[0].mine_work_status_code
+        return "UNKNOWN"
```
Comment (mhuseinov) [inline services/core-api/app/api/mines/mine/models/mine.py:184]: "UNKNOWN" is not a valid status for BE. Is there any reason not to return None and handle it on FE as "UNKNOWN" or something else if needed?

File: services/core-api/app/api/mines/work_information/models/mine_work_information.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        return f'<{self.__class__.__name__} {self.mine_work_information_id}>'
+
+    def save(self):
+        self.updated_by = User().get_user_username()
+        self.updated_timestamp = datetime.utcnow()
+        self.mine_work_status_code = self.determine_mine_work_status()
+        super(MineWorkInformation, self).save()
+
+    def determine_mine_work_status(self):
+        today = datetime.now(tz=timezone('US/Pacific')).date()
```
Comment (mhuseinov) [inline services/core-api/app/api/mines/work_information/models/mine_work_information.py:46]: UTC?

File: services/core-web/src/styles/components/CoreTooltip.scss [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.core-tooltip,
+.core-tooltip * {
+  color: white !important;
```
Comment (mhuseinov) [inline services/core-web/src/styles/components/CoreTooltip.scss:3]: this one is risky 😆

**Review by mhuseinov — COMMENTED** [review]
Could you include action creators tests? 
Also, I have not noticed any snapshot tests or their updates.

File: migrations/sql/V2021.05.25.15.10__create_mine_work_information_table.sql (hunk 2 above)
Comment (exzizt) [reply migrations/sql/V2021.05.25.15.10__create_mine_work_information_table.sql:2]: No reason, I guess it was an oversight. I could change them to be 3 chars only.

File: services/core-api/app/api/mines/work_information/models/mine_work_information.py (hunk 4 above)
Comment (exzizt) [reply services/core-api/app/api/mines/work_information/models/mine_work_information.py:46]: When users set the start/stop dates they are setting the date in pacific time so we should grab "today" as the current date in pacific time and not UTC.

Comment (mhuseinov) [reply services/core-api/app/api/mines/work_information/models/mine_work_information.py:46]: The BE and should not be tied to a user's location, it is FE responsibility. FE can send timezone on convert time to UTC so BE will be able to handle it properly no matter what region request is coming from

**Comment by exzizt** [comment]
> Could you include action creators tests?
> Also, I have not noticed any snapshot tests or their updates.

I was hoping no one would notice 😉 Kidding
Added them.

File: services/core-api/app/api/mines/work_information/models/mine_work_information.py (hunk 4 above)
Comment (exzizt) [reply services/core-api/app/api/mines/work_information/models/mine_work_information.py:46]: I actually have to revert a lot of my code back to my original implementation. I can't remember why exactly I changed it but the work status has to be a hybrid property and not a normal property. Will talk about it in the morning

File: services/core-api/app/api/mines/mine/models/mine.py [hunk 6]
```diff
@@ -1,18 +1,19 @@
...
+from sqlalchemy import func, literal, select, desc, and_
+
 from geoalchemy2 import Geometry
 from app.extensions import db
 from app.api.utils.models_mixins import SoftDeleteMixin, AuditMixin, Base
 from app.api.mines.permits.permit.models.permit import Permit
 from app.api.mines.permits.permit.models.mine_permit_xref import MinePermitXref
 from app.api.users.minespace.models.minespace_user_mine import MinespaceUserMine
-from app.api.mines.government_agencies.models.government_agency_type import GovernmentAgencyType
+from app.api.mines.work_information.models.mine_work_information import MineWorkInformation
```
Comment (mmumby) [inline services/core-api/app/api/mines/mine/models/mine.py:16]: Was this deletion intentional?

File: services/core-web/src/components/mine/Summary/MineWorkInformation.js [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+  updateMineWorkInformation: PropTypes.func.isRequired,
+  deleteMineWorkInformation: PropTypes.func.isRequired,
+  openModal: PropTypes.func.isRequired,
+  closeModal: PropTypes.func.isRequired,
+};
+
+const defaultProps = {};
+
+const { Text } = Typography;
+
```
Comment (mmumby) [inline services/core-web/src/components/mine/Summary/MineWorkInformation.js:39]: Is this syntax passing unit tests? The antd4 updates had issues with referencing components this way

**Review by mmumby — APPROVED** [review]
gave it a quick review since Mekhti already revied. I have a few comments but nothing blocking. Well done, Luke!

File: services/core-web/src/components/mine/Summary/MineWorkInformation.js (hunk 7 above)
Comment (exzizt) [reply services/core-web/src/components/mine/Summary/MineWorkInformation.js]: NO it wasn't. Fixed

File: services/core-api/app/api/mines/mine/models/mine.py (hunk 6 above)
Comment (exzizt) [reply services/core-api/app/api/mines/mine/models/mine.py:16]: Yup, it wasn't being used.

**Review by mmumby — APPROVED** [review]

**Pull request merged**